### PR TITLE
Balanced_epcr_qc

### PR DESCRIPTION
**Pull request by kenissur**
This PR contains a few new functions:
VcfAnnotator 
This is a framework for annotating vcf files, including FeatureMaps, and that is the reason why I want to merge this now - I want to incorporate that into Rina's PR

read_sorter_statistics_csv
Auxiliary function to read the csv file Sorter produces and is uploaded along with the cram file (or from trim_align_sort)

balanced_strand_utils
A set of functions for balanced strand (balanced ePCR) QC, all tested and working but still require a main function and a report
This function is almost 1000 lines so you could skip reviewing that thoroughly I guess

### CODE REVIEW

File: ugvc/mrd/balanced_strand_utils.py [hunk 1]
```diff
@@ -0,0 +1,992 @@
...
+                    )  # this works for nan values as well - returns UNDETERMINED
+                print(
+                    is_end_reached,
+                    balanced_tags[TrimmerSegmentTags.NATIVE_ADAPTER.value],
+                    balanced_tags[TrimmerSegmentTags.STEM_END.value],
+                    tags_sum_end,
+                    (self.min_total_hmer_lengths_in_tags <= tags_sum_end <= self.max_total_hmer_lengths_in_tags),
+                    balanced_tags,
+                )
+                print(record)
```
Comment (GatKrieger) [inline ugvc/mrd/balanced_strand_utils.py]: remove print

File: ugvc/mrd/balanced_strand_utils.py [hunk 2]
```diff
@@ -0,0 +1,992 @@
...
+    return df_category_concordance
+
+
+def collect_statistics(
+    adapter_version: str | BalancedStrandAdapterVersions,
+    trimmer_histogram_csv: str,
+    sorter_stats_csv: str,
+    output_filename: str,
+    input_material_ng: float = None,
+) -> pd.DataFrame:
```
Comment (GatKrieger) [inline ugvc/mrd/balanced_strand_utils.py]: does not return anything

File: ugvc/mrd/balanced_strand_utils.py [hunk 3]
```diff
@@ -0,0 +1,992 @@
...
+                    balanced_tags,
+                )
+                print(record)
+            records_out[j] = record
+
+        return records_out
+
+
+# Display defaults
+STRAND_RATIO_AXIS_LABEL = "LIG/HYB strands ratio"
```
Comment (doron-st) [inline ugvc/mrd/balanced_strand_utils.py]: maybe move to top of file?

File: ugvc/mrd/balanced_strand_utils.py [hunk 4]
```diff
@@ -0,0 +1,992 @@
...
+
+# Misc
+balanced_category_list = [v.value for v in BalancedCategories.__members__.values()]
+supported_adapter_versions = BalancedStrandAdapterVersions.__members__.values()
```
Comment (doron-st) [inline ugvc/mrd/balanced_strand_utils.py:226]: ditto

Comment (doron-st) [reply ugvc/mrd/balanced_strand_utils.py:226]: and capitalize for constant?

File: ugvc/mrd/balanced_strand_utils.py [hunk 5]
```diff
@@ -0,0 +1,992 @@
...
+            + f"supprted values are {', '.join(supported_adapter_versions)}"
+        )
+    if isinstance(adapter_version, str):
+        assert adapter_version in [v.value for v in adapter_version], (
+            f"Unsupported adapter version {adapter_version}, "
+            + f"supprted values are {', '.join(supported_adapter_versions)}"
+        )
+
+
+def get_annotation(x, sr_lower, sr_upper):
```
Comment (doron-st) [inline ugvc/mrd/balanced_strand_utils.py]: will be nice to rename x is strand_ratio.
function name is also a bit too generic.

**Review by doron-st — COMMENTED** [review]
Looks very good. minor comments

File: ugvc/mrd/balanced_strand_utils.py [hunk 6]
```diff
@@ -0,0 +1,992 @@
...
+    return ax
+
+
+def plot_strand_ratio_category_concordnace(
+    adapter_version: str | BalancedStrandAdapterVersions,
+    df_trimmer_histogram: pd.DataFrame,
+    title: str = "",
+    output_filename: str = None,
+    axs: list[plt.Axes] = None,
+) -> list[plt.Axes]:
```
Comment (GatKrieger) [inline ugvc/mrd/balanced_strand_utils.py:903]: also this one does not return anything

**Review by GatKrieger — COMMENTED** [review]
Thanks for going through this with me!
how do the plots look like?

File: ugvc/mrd/balanced_strand_utils.py (hunk 3 above)
Comment (kenissur) [reply ugvc/mrd/balanced_strand_utils.py]: Done

File: ugvc/mrd/balanced_strand_utils.py (hunk 4 above)
Comment (kenissur) [reply ugvc/mrd/balanced_strand_utils.py:226]: They are defined by the classes above so they can't be in the top of the file. They're not exactly contants either, not sure what the right convention is...

File: ugvc/mrd/balanced_strand_utils.py (hunk 5 above)
Comment (kenissur) [reply ugvc/mrd/balanced_strand_utils.py]: Done

File: ugvc/mrd/balanced_strand_utils.py (hunk 6 above)
Comment (kenissur) [reply ugvc/mrd/balanced_strand_utils.py:903]: Fixed

**Review by GatKrieger — APPROVED** [review]

**Review by rinashalibo — COMMENTED** [review]
looks good! perhaps its worth adding am example code for the abstract methods - or direct to a reference

**Pull request merged**